### PR TITLE
E2E test for trust domain alias client side secure naming.

### PR DESCRIPTION
**Pull request by liwenhao0810**
Please provide a description for what this PR is for.

And to help us figure out who should review this PR, please 
put an X in all the areas that this PR affects.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[x] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any changes that may affect Istio users.

### CODE REVIEW

**Comment by liwenhao0810** [comment]
@incfly

File: tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+	POLICY = `
+apiVersion: "security.istio.io/v1beta1"
+kind: "PeerAuthentication"
+metadata:
+  name: "mtls"
+spec:
+  mtls:
+    mode: STRICT
+---
+apiVersion: "networking.istio.io/v1alpha3"
```
Comment (incfly) [inline tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go:33]: I believe this is not needed. Have you tried running test without ISTIO_MUTUAL?

File: tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+	workDir := path.Join(env.IstioSrc, "samples/certs")
+	script := path.Join(workDir, "generate-workload.sh")
+	command := exec.Cmd{
+		Path:   script,
+		Args:   []string{script, td, ns, sa, workDir, "tmp"},
+		Stdout: os.Stdout,
+		Stderr: os.Stdout,
+	}
+	if err := command.Run(); err != nil {
+		ctx.Logf("Failed to create testing certificates: %s", err)
```
Comment (incfly) [inline tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go:173]: Is Errorf available?

File: tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+	}
+}
+
+func cleanCerts(ctx framework.TestContext) {
+	err := os.RemoveAll(path.Join(env.IstioSrc, "samples/certs/tmp"))
+	if err != nil {
+		ctx.Logf("Failed to clean testing certificates: %s", err)
```
Comment (incfly) [inline tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go:180]: Same here. The test should fail when hitting this line.

File: tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+
+func loadCert(t test.Failer, name string) string {
+	data, err := cert.ReadSampleCertFromFile(name)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return string(data)
+}
+
+func generateCerts(ctx framework.TestContext, ns string, td string, sa string) {
```
Comment (incfly) [inline tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go:163]: why test.Failer above and context here? you only doing a loging operation?

File: tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+	workDir := path.Join(env.IstioSrc, "samples/certs")
+	script := path.Join(workDir, "generate-workload.sh")
+	command := exec.Cmd{
+		Path:   script,
+		Args:   []string{script, td, ns, sa, workDir, "tmp"},
```
Comment (incfly) [inline tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go:168]: You can make cleanup function as a returned value. and the caller would just doing 

```
cleanup := generateCerts()
defer cleanup()
```

This makes logic tigher. e.g. you don't have to remember `tmp` folder is copied in two places.

File: tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+						RootCert:   loadCert(t, "root-cert.pem"),
+						ClientCert: loadCert(t, "tmp/workload-bar-cert.pem"),
+						Key:        loadCert(t, "tmp/workload-bar-key.pem"),
+					},
+				}).
+				BuildOrFail(t)
+
+			ctx.Config().ApplyYAMLOrFail(ctx, testNS.Name(), fmt.Sprintf(POLICY))
+
+			verify := func(t *testing.T, src echo.Instance, dest echo.Instance, s scheme.Instance, allow bool) {
```
Comment (incfly) [inline tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go:126]: allow/deny naming is inherited from validatoin test which make sense since it's testing server side behavior. Here you are not. (ALLOW/DENY is server side terminology)

Change to `success`

File: tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+// TestTrustDomainAliasClient scope:
+// The client side mTLS connection should validate the trust domain alias during secure naming validation.
+// - When server side has trust domain which defined as trust domain alias, the mTLS connection should expect success
+// - When server side has trust domain which not defined as trust domain alias, the mTLS connection should expect fail
+func TestTrustDomainAliasClient(t *testing.T) {
+	framework.NewTest(t).
+		Features("security.peer.trust-domain-validation-alias-client").
+		Run(func(ctx framework.TestContext) {
+			testNS := namespace.NewOrFail(t, ctx, namespace.Config{
+				Prefix: "trust-domain-alias-client-validation",
```
Comment (incfly) [inline tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go:54]: change namespace to `xx-alias`. not validatoin

File: tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+  host: "*.local"
+  trafficPolicy:
+    tls:
+      mode: ISTIO_MUTUAL
+`
+)
+
+// TestTrustDomainAliasClient scope:
+// The client side mTLS connection should validate the trust domain alias during secure naming validation.
+// - When server side has trust domain which defined as trust domain alias, the mTLS connection should expect success
```
Comment (incfly) [inline tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go:47]: This sentence does not read flutently without context. Changed to something like,

The trust domain of server TLS certificate is not in the list of MeshConfig.TrustDomainAliases, then ...

Comment (incfly) [inline tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go:47]: expect is unnecessary, just `should succeed.` same above

File: tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+					Subsets: []echo.SubsetConfig{
+						{
+							Annotations: echo.NewAnnotations().SetBool(echo.SidecarInject, false),
+						},
+					},
+					ServiceAccount: true,
+					Ports: []echo.Port{
+						{
+							Name:         HTTPS,
+							Protocol:     protocol.HTTPS,
```
Comment (incfly) [inline tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go:86]: Do you have to define your own port? Are there any default Echo config with 443 port we can use?

File: tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+			//generateCerts(ctx, testNS.Name())
+			generateCerts(ctx, testNS.Name(), "foo", "server-naked-foo")
+			generateCerts(ctx, testNS.Name(), "bar", "server-naked-bar")
+			defer cleanCerts(ctx)
+
+			// Deploy 3 workloads:
+			// client: echo app with istio-proxy sidecar injected, holds default trust domain cluster.local
+			// serverNakedFoo: echo app without istio-proxy sidecar, holds custom trust domain trust-domain-foo
+			// serverNakedBar: echo app without istio-proxy sidecar, holds custom trust domain trust-domain-bar
+			var client, serverNakedFoo, serverNakedBar echo.Instance
```
Comment (incfly) [inline tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go:68]: when running your test i remember we see two client, one client-b one client-v1. we should only need one client workload. Could you confirm?

File: tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+  name: "server-naked"
+spec:
+  host: "*.local"
+  trafficPolicy:
+    tls:
+      mode: ISTIO_MUTUAL
+`
+)
+
+// TestTrustDomainAliasClient scope:
```
Comment (incfly) [inline tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go:45]: This test's setup requires much more comment for context. In order to make this test working, we 

1. part of custom root? Why?
1. Generate customized certificate for server workload.
1. Server workloads are naked.
1. Expected behavior with given config.

It should be obvious after reading comments & code, why each of above are needed for this test.

File: tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go (hunk 4 above)
Comment (incfly) [inline tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go:163]: not needed. `ns, td, sa string` in golang

File: tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+					checker := connection.Checker{
+						From:          src,
+						Options:       opt,
+						ExpectSuccess: allow,
+					}
+					checker.CheckOrFail(ctx)
+				})
+			}
+
+			verify(t, client, serverNakedFoo, scheme.HTTP, true)
```
Comment (incfly) [inline tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go:150]: Use table driven test. rather sequential ordering.

**Review by incfly — COMMENTED** [review]
I remember there's file for setup meshconfig somewhere in separate go file?

File: pkg/test/framework/features/features.yaml [hunk 13]
```diff
@@ -64,6 +64,7 @@ features:
     peer:
       secure-naming:
       trust-domain-validation:
+      trust-domain-validation-alias-client:
```
Comment (incfly) [inline pkg/test/framework/features/features.yaml:67]: trust-domain-alias-secure-naming  might be a better name.

File: samples/certs/generate-workload.sh [hunk 14]
```diff
@@ -15,17 +15,26 @@
 #   limitations under the License.
 
 name=${1:-foo}
-san="spiffe://trust-domain-$name/ns/$name/sa/$name"
+ns=${2:-$name}
+sa=${3:-$name}
+dir=${4:-"."}
+tmp=${5:-""}
+san="spiffe://trust-domain-$name/ns/$ns/sa/$sa"
```
Comment (incfly) [inline samples/certs/generate-workload.sh:22]: Print out an "echo" for SAN for potential debugging?

File: samples/certs/generate-workload.sh [hunk 15]
```diff
@@ -15,17 +15,26 @@
 #   limitations under the License.
 
 name=${1:-foo}
-san="spiffe://trust-domain-$name/ns/$name/sa/$name"
+ns=${2:-$name}
+sa=${3:-$name}
+dir=${4:-"."}
+tmp=${5:-""}
```
Comment (incfly) [inline samples/certs/generate-workload.sh:21]: why we need both dir and tmp? can we just use one directory?

File: tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go (hunk 4 above)
Comment (incfly) [reply tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go:163]: and the order should td, ns, sa

File: samples/certs/generate-workload.sh (hunk 15 above)
Comment (liwenhao0810) [reply samples/certs/generate-workload.sh:21]: the tmp is for the testing certs, since it uses the foo and bar, which would overwrite the existing certs.

File: pkg/test/framework/features/features.yaml (hunk 13 above)
Comment (liwenhao0810) [reply pkg/test/framework/features/features.yaml:67]: will do

File: samples/certs/generate-workload.sh (hunk 14 above)
Comment (liwenhao0810) [reply samples/certs/generate-workload.sh:22]: echo in line 54

File: tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go (hunk 1 above)
Comment (liwenhao0810) [reply tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go:33]: I didnt, will try and add here.

File: tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go (hunk 10 above)
Comment (liwenhao0810) [reply tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go:68]: have tested that one client is working as expected

File: tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go (hunk 1 above)
Comment (liwenhao0810) [reply tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go:33]: getting error without the destination rule.

**Comment by incfly** [comment]
Oh I remembered it's because without DestinationRule. Auto mTLS will send
plaintext traffic to the server side. This should be commented around the
config. Something like:

```
// We explicitly configure the client side to send mTLS traffic via
Destination Rule.
// The server applications are naked without sidecar injection.
// By default, auto mTLS makes clients send plain text traffic for
endpoints without sidecar.
```

File: tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go (hunk 3 above)
Comment (liwenhao0810) [reply tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go]: t.Fatal

File: tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go (hunk 2 above)
Comment (liwenhao0810) [reply tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go]: changed to t.Fatal

File: tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go (hunk 4 above)
Comment (liwenhao0810) [reply tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go]: changed to t.Fatal

File: tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go (hunk 5 above)
Comment (liwenhao0810) [reply tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go]: updated

File: tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go (hunk 6 above)
Comment (liwenhao0810) [reply tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go]: fixed

File: tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go (hunk 8 above)
Comment (liwenhao0810) [reply tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go]: updated comments

File: tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go (hunk 9 above)
Comment (liwenhao0810) [reply tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go]: tried without and use the auto generated port names but getting port can't find error.

File: tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go (hunk 11 above)
Comment (liwenhao0810) [reply tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go]: updated comments

File: tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go (hunk 12 above)
Comment (liwenhao0810) [reply tests/integration/security/ca_custom_root/trust_domain_alias_client_test.go]: updated.

File: samples/certs/generate-workload.sh (hunk 15 above)
Comment (incfly) [reply samples/certs/generate-workload.sh:21]: you can use tmp=$(mktmp). This is used more common in automation.

Also currently you are exposing 5 positional arguments to this shell script. It's unlikely a dev to memorize and actual fill 5 value sin the CLI.

Just assign name ot ns/sa. dir you can get it via some bash onlines, like `$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )`.  See other shell for e.g.

File: samples/certs/generate-workload.sh [hunk 16]
```diff
@@ -15,17 +15,26 @@
...
 
-openssl genrsa -out "workload-$name-key.pem" 2048
+if [ ! -d "$dir/$tmp" ]; then
+  mkdir "$dir/$tmp"
+fi
 
-cat > workload.cfg <<EOF
+openssl genrsa -out "$dir/$tmp/workload-$name-key.pem" 2048
+
+cat > "$dir"/workload.cfg <<EOF
```
Comment (incfly) [inline samples/certs/generate-workload.sh:30]: this should be under TMP_DIR. to be consistent with others.

File: samples/certs/generate-workload.sh [hunk 17]
```diff
@@ -35,15 +44,21 @@ subjectAltName = critical, @alt_names
...
-cat cert-chain.pem >> "workload-$name-cert.pem"
+cat "$dir"/cert-chain.pem >> "$dir/$tmp/workload-$name-cert.pem"
 
 echo "Generated workload-$name-[cert|key].pem with URI SAN $san"
-openssl verify -CAfile <(cat cert-chain.pem root-cert.pem) "workload-$name-cert.pem"
+openssl verify -CAfile <(cat "$dir"/cert-chain.pem "$dir"/root-cert.pem) "$dir/$tmp/workload-$name-cert.pem"
 
 # clean temporary files
-rm ca-cert.srl workload.cfg workload.csr
+if [ -f "$dir"/.srl ]; then
```
Comment (incfly) [inline samples/certs/generate-workload.sh:58]: Just need rm -rf ${TMP_DIR} if all generated files are in tmp dir.

File: tests/integration/security/ca_custom_root/trust_domain_alias_secure_naming_test.go [hunk 18]
```diff
@@ -0,0 +1,242 @@
...
+`
+)
+
+// TestTrustDomainAliasSecureNaming scope:
+// The client side mTLS connection should validate the trust domain alias during secure naming validation.
+//
+// Setup:
+// 1. Setup Istio with custom CA cert, to make sure all workloads have same root cert because of #3
+// 2. One client workload with istio-proxy sidecar injected
+// 3. Two naked server workloads with custom certs in two different trust domain reflected in the SAN (spiffe)
```
Comment (incfly) [inline tests/integration/security/ca_custom_root/trust_domain_alias_secure_naming_test.go:66]: s/in two different trust domain reflected in the SAN (spiffe)/ to 

`whose URI SAN have different SPIFFE trust domains".

File: tests/integration/security/ca_custom_root/trust_domain_alias_secure_naming_test.go [hunk 19]
```diff
@@ -0,0 +1,242 @@
...
+      mode: ISTIO_MUTUAL
+`
+)
+
+// TestTrustDomainAliasSecureNaming scope:
+// The client side mTLS connection should validate the trust domain alias during secure naming validation.
+//
+// Setup:
+// 1. Setup Istio with custom CA cert, to make sure all workloads have same root cert because of #3
+// 2. One client workload with istio-proxy sidecar injected
```
Comment (incfly) [inline tests/integration/security/ca_custom_root/trust_domain_alias_secure_naming_test.go:65]: no need mention istio-proxy. sidecar is enough

File: tests/integration/security/ca_custom_root/trust_domain_alias_secure_naming_test.go [hunk 20]
```diff
@@ -0,0 +1,242 @@
...
+    tls:
+      mode: ISTIO_MUTUAL
+`
+)
+
+// TestTrustDomainAliasSecureNaming scope:
+// The client side mTLS connection should validate the trust domain alias during secure naming validation.
+//
+// Setup:
+// 1. Setup Istio with custom CA cert, to make sure all workloads have same root cert because of #3
```
Comment (incfly) [inline tests/integration/security/ca_custom_root/trust_domain_alias_secure_naming_test.go:64]: This is a bit hard to follow, change to:

```
Setup Istio with custom CA cert. This is because we need to use that root cert to sign customized certificate for server workloads to give them different trust domains.
```

File: tests/integration/security/ca_custom_root/trust_domain_alias_secure_naming_test.go [hunk 21]
```diff
@@ -0,0 +1,242 @@
...
+// 2. One client workload with istio-proxy sidecar injected
+// 3. Two naked server workloads with custom certs in two different trust domain reflected in the SAN (spiffe)
+// 4. PeerAuthentication with strict mtls, to enforce the mtls connection
+// 5. DestinaitonRule with tls ISTIO_MUTUAL mode, because Istio auto mTLS will let client sends plaintext to naked servers by default
```
Comment (incfly) [inline tests/integration/security/ca_custom_root/trust_domain_alias_secure_naming_test.go:68]: s/sends/send.

File: tests/integration/security/ca_custom_root/trust_domain_alias_secure_naming_test.go [hunk 22]
```diff
@@ -0,0 +1,242 @@
...
+// 6. MeshConfig.TrustDomainAliases contains one of the trust domain in #3
```
Comment (incfly) [inline tests/integration/security/ca_custom_root/trust_domain_alias_secure_naming_test.go:69]: Be explicit about the values. Aliases configured as "server-naked-foo". in #3 we didn't mention values either. makes hard to udnerstand.

File: tests/integration/security/ca_custom_root/trust_domain_alias_secure_naming_test.go [hunk 23]
```diff
@@ -0,0 +1,242 @@
...
+//
+// Expectation:
+// 1. When the trust domain of server TLS certificate is in the list of MeshConfig.TrustDomainAliases,
```
Comment (incfly) [inline tests/integration/security/ca_custom_root/trust_domain_alias_secure_naming_test.go:72]: could be simplified by: client requests to "server-naked-foo" succeeds, and requests to "server-naked-bar" fails.

Or if you want to keep current. 2) can be removed by: "Otherwise, requests fail."

File: tests/integration/security/ca_custom_root/trust_domain_alias_secure_naming_test.go [hunk 24]
```diff
@@ -0,0 +1,242 @@
...
+func TestTrustDomainAliasSecureNaming(t *testing.T) {
+	framework.NewTest(t).
+		Features("security.peer.trust-domain-alias-secure-naming").
+		Run(func(ctx framework.TestContext) {
+			testNS := namespace.NewOrFail(t, ctx, namespace.Config{
+				Prefix: "trust-domain-alias",
+				Inject: true,
+			})
+
+			// Create testing certs using runtime namespace
```
Comment (incfly) [inline tests/integration/security/ca_custom_root/trust_domain_alias_secure_naming_test.go:85]: All comments end up with "." to be a complete setence. (through out the PR)

File: tests/integration/security/ca_custom_root/trust_domain_alias_secure_naming_test.go [hunk 25]
```diff
@@ -0,0 +1,242 @@
...
+			cleanup := generateCerts(ctx, testNS.Name())
+			defer cleanup()
+
+			// Deploy 3 workloads:
+			// client: echo app with istio-proxy sidecar injected, holds default trust domain cluster.local
+			// serverNakedFoo: echo app without istio-proxy sidecar, holds custom trust domain trust-domain-foo
```
Comment (incfly) [inline tests/integration/security/ca_custom_root/trust_domain_alias_secure_naming_test.go:91]: same as above

File: samples/certs/generate-workload.sh (hunk 15 above)
Comment (incfly) [reply samples/certs/generate-workload.sh:21]: Ignore the `tmp` one i realized you need that as argument to invoke from golang code. other arguments i think can still be reduced.

File: tests/integration/security/ca_custom_root/trust_domain_alias_secure_naming_test.go [hunk 26]
```diff
@@ -0,0 +1,242 @@
...
+		{
+			td: "bar",
+			sa: "server-naked-bar",
+		},
+	}
+
+	for _, crt := range crts {
+		command := exec.Cmd{
+			Path:   script,
+			Args:   []string{script, crt.td, ns, crt.sa, workDir, "tmp"},
```
Comment (incfly) [inline tests/integration/security/ca_custom_root/trust_domain_alias_secure_naming_test.go:227]: use a constant var to refer to it? This hardcoded value is used multiple times in the code?

File: tests/integration/security/ca_custom_root/trust_domain_alias_secure_naming_test.go [hunk 27]
```diff
@@ -0,0 +1,242 @@
...
+// 1. Setup Istio with custom CA cert, to make sure all workloads have same root cert because of #3
+// 2. One client workload with istio-proxy sidecar injected
+// 3. Two naked server workloads with custom certs in two different trust domain reflected in the SAN (spiffe)
+// 4. PeerAuthentication with strict mtls, to enforce the mtls connection
+// 5. DestinaitonRule with tls ISTIO_MUTUAL mode, because Istio auto mTLS will let client sends plaintext to naked servers by default
+// 6. MeshConfig.TrustDomainAliases contains one of the trust domain in #3
+//
+// Expectation:
+// 1. When the trust domain of server TLS certificate is in the list of MeshConfig.TrustDomainAliases,
+//    the connection from client to server should succeed.
```
Comment (incfly) [inline tests/integration/security/ca_custom_root/trust_domain_alias_secure_naming_test.go:73]: the mTLS connection.
same below

File: tests/integration/security/ca_custom_root/trust_domain_alias_secure_naming_test.go [hunk 28]
```diff
@@ -0,0 +1,242 @@
...
+		}
+		if err := command.Run(); err != nil {
+			t.Fatal("Failed to create testing certificates: %s", err)
+		}
+	}
+
+	return func() {
+		err := os.RemoveAll(path.Join(env.IstioSrc, "samples/certs/tmp"))
+		if err != nil {
+			t.Fatal("Failed to clean testing certificates: %s", err)
```
Comment (incfly) [inline tests/integration/security/ca_custom_root/trust_domain_alias_secure_naming_test.go:239]: clean/clean up

**Review by incfly — COMMENTED** [review]
looks good overall. mostly just some comments clarification
/lgtm

**Comment by incfly** [comment]
@ericvn mind taking a look as wg test-release/doc owner? Thanks!

File: samples/certs/generate-workload.sh (hunk 15 above)
Comment (liwenhao0810) [reply samples/certs/generate-workload.sh]: removed dir as argument

File: samples/certs/generate-workload.sh (hunk 17 above)
Comment (liwenhao0810) [reply samples/certs/generate-workload.sh]: The tmp dir only contains the temp certs, on purposely keep other files still in current dir which not impacts original script. And the tmp dir would need to be cleaned up in go code after the test execution.

File: tests/integration/security/ca_custom_root/trust_domain_alias_secure_naming_test.go (hunk 18 above)
Comment (liwenhao0810) [reply tests/integration/security/ca_custom_root/trust_domain_alias_secure_naming_test.go]: fixed

File: samples/certs/generate-workload.sh (hunk 16 above)
Comment (liwenhao0810) [reply samples/certs/generate-workload.sh]: The tmp dir only contains the temp certs, on purposely keep other files still in current dir which not impacts original script

**Comment by incfly** [comment]
@istio/wg-docs-maintainers @istio/wg-test-and-release-maintainers could you take a look of this PR? Thanks.

**Review by ericvn — APPROVED** [review]

**Review by incfly — APPROVED** [review]
